[Winston Cahya] iP

CommanderW324 opened the pull request:
# Duke ChatBot

DukePro frees your mind of having to remember things you need to do. It's,

* text-based
* easy to learn
* Uncomplicated
As (Probably) Sun Tzu said:

> Don't complicate matters

All you need to do is,

1. download it from [here](http://github.com/CommanderW324/ip).
1. double-click it.
1. add your tasks.
1. let it manage your tasks for you 😀 
- [x] Easy to use
- [ ] Joyful
Practice Java : 
``` java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
 }
```
`<Easy !>`
~~Difficulty~~ with this app

Commit:
Add Gradle support

CommanderW324 committed:
Level- 1

CommanderW324 committed:
Fixed Level-1

CommanderW324 committed:
Level-2

CommanderW324 committed:
Level-3

CommanderW324 committed:
Level-4

CommanderW324 committed:
Level-4 Completed

CommanderW324 committed:
Text-UI Testing

CommanderW324 committed:
Level-5

CommanderW324 committed:
Level-6 AND error Handling added

CommanderW324 committed:
Level-7

CommanderW324 committed:
Level-8

CommanderW324 committed:
Level-7Revised

CommanderW324 committed:
Added LocalDates

CommanderW324 committed:
Merge branch 'branch-Level-7' into master

CommanderW324 committed:
Resolved Merging conflicts

jeromewjj commented on data/DukeTask.txt:
```diff
@@ -0,0 +1,5 @@
+
+[T][ ]          kill humans
+[T][ ]          kill pokemon
```
Erhmmm, I hope it will never be mark done😰  Should have added this into .gitignore.

jeromewjj commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,189 @@
+                        (x,y) -> x + "\n" + y);
+                Files.write(taskListPath, tasksinstring.getBytes());
+            } catch(IOException e) {
+                System.out.println(e.getMessage());
```
The function seems too long. Could split the function into the different functions to handle different parts.

jeromewjj commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,189 @@
+            }
+        }
+    }
+    public static Task handleInput(String userInput) throws DukeException{
```
Could add a space after DukeException

GWphua replied:
Would be better if 139 is left as a blank line, and this method signature is written on the 140th line instead.
Try `DukeException {` instead of `DukeException{` to better adhere to the style guide.

jeromewjj commented on src/main/java/TestOut.java:
```diff
@@ -0,0 +1,43 @@
+            }
+        }
+        try {
+//            FileWriter fw = new FileWriter(taskListPath.normalize().toString());
```
Could be deleted since it is no longer being used

jeromewjj reviewed (commented):
The Duke class method seems to be too long and there were a few violations of the Java coding standard. Maybe keep your intention hidden next time. 🤣

GWphua commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,189 @@
+                    Character isDone = curr.charAt(4);
+                    String desc = curr.substring(6);
+                    boolean done = isDone == 'X';
+                    if(tasktype == 'T') {
+                        currTask = new Todo(desc, done);
+                    } else if(tasktype == 'E') {
+                        int index = desc.indexOf("at:");
+                        int n = desc.length();
+                        String date = desc.substring(index + 4, n-1);
+                        currTask = new Event(desc.substring(0,index), done, LocalDate.parse(date));
+                    } else if(tasktype == 'D') {
+                        int index = desc.indexOf("by:");
+                        int n = desc.length();
+                        String date = desc.substring(index + 4, n-1);
+
+                        currTask = new Deadline(desc.substring(0, index), done, LocalDate.parse(date));
+                    }
```
Perhaps using switch statements would be a more elegant way of writing this?

GWphua commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,189 @@
+                }
+                System.out.println("\n ----------------------------------");
+            }
+            try{
```
Using `try {` would better adhere to the Java Style.

GWphua commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,189 @@
 public class Duke {
+    private static ArrayList<Task> tasks = new ArrayList<>();
+
+    public static void talk() {
```
Perhaps creating new classes (Storage/UI) would help to abstract this huge line of code?

GWphua commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,189 @@
+            try{
+                Files.createDirectory(dirpath);
+            } catch (IOException e) {
+                System.out.println(e.getMessage());
```
Maybe can write these functions as DukeExceptions instead, so that all these errors can be abstracted for a method "printDukeExceptionMessage".

GWphua commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,189 @@
+            }
+        } else {
+            try {
+                File taskList = new File(taskListPath.toString());
```
Maybe try creating a new class to represent taskList?

GWphua commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,189 @@
+        } else {
+            throw new DukeException("I don't understand what you are talking about !");
+        }
+    }
```
I like the way you handle string manipulations!! Learnt a lot from these 👍

GWphua reviewed (commented):
A few nibs on Java Style, may need more OOP in the code to make it more elegant. Other than that, LTGM.

JoelChanZhiYang commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,189 @@
+import java.nio.file.Files;
+import java.io.File;
+import java.io.FileWriter;
+import java.util.stream.*;
```
Maybe you could make the ordering of your import statement more consistent?

JoelChanZhiYang commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,189 @@
+                    System.out.println("\n ----------------------------------");
+                }
+
+            } else if (userInput.startsWith("delete")) {
```
I like the way you use your `startsWith` function

JoelChanZhiYang commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,189 @@
+                throw new DukeException("You need to specify at using /by !");
+            }
+            if(task.replaceAll("\\s+","").equals("")){
+                //if remaining string is whitespace or empty
```
Good comments! It really helped me understand the code better.

JoelChanZhiYang commented on src/main/java/TestOut.java:
```diff
@@ -0,0 +1,43 @@
+import java.io.IOException;
+import java.nio.file.Paths;
+import java.nio.file.Files;
+import java.io.FileWriter;
```
Your imports would adhere more to the coding standards if they were more consistent!

JoelChanZhiYang reviewed (commented):
I really like the way you handle your strings. Your code would be more readable if you abstracted some of the logic into functions and classes.

CommanderW324 committed:
Refactored Parser and Storage, fixed bug of adding whitespace everytime loading task.

CommanderW324 committed:
Refactored Classes

CommanderW324 committed:
Package organization

CommanderW324 committed:
Added JUNIT Test

CommanderW324 committed:
Created jar files

CommanderW324 committed:
Added Javadocs

CommanderW324 committed:
Matched code to coding standard

CommanderW324 committed:
Added finding keyword in the tasklist

CommanderW324 committed:
Merge branch 'branch-A-CodingStandard' into master

CommanderW324 committed:
Merge branch 'branch-Level-9' into master

CommanderW324 committed:
Fixed conflicting UI message

CommanderW324 committed:
Merge branch 'add-gradle-support' into master

CommanderW324 committed:
Started on GUI

CommanderW324 committed:
GUI for Duke has been added

CommanderW324 committed:
Created fat Jar File

CommanderW324 committed:
Added assertions to the Parser and Storage Class

Assertions are used to verify assumptions of the program so that these
assumptions are valid. This will allow error to be notified during
development

CommanderW324 committed:
Abstracted Magic Numbers and Improved Code Quality

To improve quality of Code I added constant declarations to remove magic
numbers in Parser and Storage Class. I also deleted the code for the
Command Line Version of the Duke and did deletion of unused codes.

CommanderW324 committed:
Fixed constructor for Duke

CommanderW324 committed:
Merge pull request #2 from CommanderW324/branch-A-Assertions

Added assertions to the Parser and Storage Class

CommanderW324 committed:
Merge pull request #3 from CommanderW324/branch-A-CodeQuality

Branch a code quality

CommanderW324 committed:
Added sorting functionality for the tasks

Todo tasks will be sorted at the first according to their order of
entering. Event and deadline will be sorted chronologically according to
the date.

CommanderW324 committed:
Acknowledgements to reused code

CommanderW324 committed:
Merge pull request #4 from CommanderW324/branch-C-Sort

Added sorting functionality for the tasks

CommanderW324 committed:
Added Ui.png

CommanderW324 committed:
UserGuide Creation

CommanderW324 committed:
Set theme jekyll-theme-slate

CommanderW324 committed:
Delete Ui.png.png

CommanderW324 committed:
Add files via upload

CommanderW324 committed:
Update README.md
